src/v/storage/segment_appender.cc

```diff
@@ -112,11 +113,15 @@ segment_appender::segment_appender(segment_appender&& o) noexcept
 ss::future<> segment_appender::append(const model::record_batch& batch) {
     _batch_types_to_write |= 1LU << static_cast<uint8_t>(batch.header().type);
 
-    auto hdrbuf = std::make_unique<iobuf>(
-      storage::batch_header_to_disk_iobuf(batch.header()));
-    auto ptr = hdrbuf.get();
-    return append(*ptr).then(
-      [this, &batch, cpy = std::move(hdrbuf)] { return append(batch.data()); });
+    auto hdrbuf = storage::batch_header_to_disk_iobuf(batch.header());
+
+    auto batch_disk_size = hdrbuf.size_bytes() + batch.data().size_bytes();
+    if (next_committed_offset() + batch_disk_size > _fallocation_offset) {
```

Comment:
maybe put this if statement into a small function as the condition is duplicated